AdvancedDLSupport/ImplementationGenerators/Wrappers/DelegateWrapper.cs

```diff
@@ -0,0 +1,221 @@
+//
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using System.Runtime.InteropServices;
+using AdvancedDLSupport.Extensions;
+using AdvancedDLSupport.Pipeline;
+using AdvancedDLSupport.Reflection;
+using JetBrains.Annotations;
+using StrictEmit;
+using static AdvancedDLSupport.ImplementationGenerators.GeneratorComplexity;
+
+namespace AdvancedDLSupport.ImplementationGenerators
+{
+    /// <summary>
+    /// Generates wrapper instructions for marshalling delegate type.
+    /// </summary>
+    public class DelegateWrapper : CallWrapperBase
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DelegateWrapper"/> class.
+        /// </summary>
+        /// <param name="targetModule">The module where the implementation should be generated.</param>
+        /// <param name="targetType">The type in which the implementation should be generated.</param>
+        /// <param name="targetTypeConstructorIL">The IL generator for the target type's constructor.</param>
+        /// <param name="options">The configuration object to use.</param>
+        public DelegateWrapper([NotNull] ModuleBuilder targetModule, [NotNull] TypeBuilder targetType, [NotNull] ILGenerator targetTypeConstructorIL, ImplementationOptions options)
+            : base(targetModule, targetType, targetTypeConstructorIL, options)
+        {
+        }
+
+        /// <inheritdoc/>
+        public override GeneratorComplexity Complexity => MemberDependent | TransformsParameters | CreatesTypes;
+
+        /// <inheritdoc />
+        public override bool IsApplicable(IntrospectiveMethodInfo member)
+        {
+            if (member.ReturnType.IsDelegate())
+            {
+                return true;
+            }
+
+            foreach (var p in member.ParameterTypes)
+            {
+                if (p.IsDelegate())
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <inheritdoc/>
+        public override void EmitPrologue(ILGenerator il, PipelineWorkUnit<IntrospectiveMethodInfo> workUnit)
+        {
+            var definition = workUnit.Definition;
+
+            var allocMethod =
+                typeof(NativeLibraryBase).GetMethod("AddLifetimeDelegate", BindingFlags.Instance | BindingFlags.NonPublic);
+            il.EmitLoadArgument(0);
+            for (short i = 1; i <= definition.ParameterTypes.Count; ++i)
+            {
+                var parameterType = definition.ParameterTypes[i - 1];
+
+                if (!parameterType.IsDelegate())
+                {
+                    il.EmitLoadArgument(i);
+                    continue;
+                }
+
+                var lifetime = GetParameterDelegateLifetime(definition.ParameterCustomAttributes[i - 1]);
+
+                if (lifetime != DelegateLifetimeAttribute.DelegateLifetime.CallOnly)
```

Comment:
This branch runs even when `lifetime` is `None`.

Reply:
Well for me None, was the "default" and I was not too sure on what was meant to be done otherwise with None(or should it be removed completly, which I would have normally done, but the enum - except name - was given by you, so what should None be according to your understanding?
But nevertheless you are right, no matter what it should be changed to check fo explicit values, for times when a new one gets added..

Reply:
In my mind, `None` implies that the user needs to manage the delegate lifetime and pinning themselves, thus it should not be cached.